ateam_vision_filter/src/filters/interacting_multiple_model_filter.hpp

```diff
@@ -0,0 +1,125 @@
+#define FILTERS__INTERACTING_MULTIPLE_MODEL_FILTER_HPP_
+
+// Multiple kalman filters each with a different model
+
+// Ball with rolling accel
+// Ball with sliding accel
+// Ball with static velocity
+// Ball with bounce velocity in estimate direction
+// Ball with slow kick robot facing direction
+// Ball with medium kick robot facing direction
+// Ball with fast kick robot facing direciton
+
+// Robot with constant accel
+
+#include <Eigen/Dense>
+
+#include <map>
+#include <memory>
+#include <vector>
+
+#include "filters/kalman_filter.hpp"
+#include "types/models.hpp"
+#include "generators/model_input_generator.hpp"
+#include "generators/transmission_probability_generator.hpp"
+
+class InteractingMultipleModelFilter
+{
+public:
+  void setup(
+    const KalmanFilter & base_model,
+    std::vector<Models::ModelType> model_types,
+    std::shared_ptr<ModelInputGenerator> model_input_generator,
+    std::shared_ptr<TransmissionProbabilityGenerator> transmission_probability_generator);
+
+  /**
+   * Clone filter and set state estimate
+   *
+   * @note Clones all current tracks and covariance estimates. Only state estimates get changed
+   * to new values
+   */
+  InteractingMultipleModelFilter clone(const Eigen::VectorXd & measurement);
+
+  /**
+   * Step models forward one time step
+   */
+  void predict();
+
+  /**
+   * Update models with new measurement
+   */
+  void update(const Eigen::VectorXd & measurement);
+
+  /**
+   * @return Best estimation of state (special weighted average of model's xhat)
+   */
+  Eigen::VectorXd get_state_estimate() const;
+
+  Eigen::VectorXd get_position_estimate() const;
+
+  /**
+   * @return True if this filter has been updated at a reasonable regularly-ness
+   */
+  bool has_been_updated_regularly() const;
+
+  /**
+   * @return Score representing how much this filter has been updated
+   */
+  double get_validity_score() const;
+
+  /**
+   * @return Score of how far off a potential measurement error is
+   */
+  double get_potential_measurement_error(const Eigen::VectorXd & measurement);
+
+private:
+  void update_mu(const Eigen::VectorXd & zt);
+
+  /**
+   * Calculates the PDF of a multivariate normal distribution
+   *
+   * @param X Test point
+   * @param mu mean
+   * @param sigma variance
+   *
+   * @return PDF at test point
+   */
+  static double normal_multivariate_distribution_pdf(
+    Eigen::VectorXd x, Eigen::VectorXd mu,
+    Eigen::MatrixXd sigma);
+
+  std::vector<Models::ModelType> model_types;  // List of models
+  std::map<Models::ModelType, KalmanFilter> models;  // Kalman filter representing ModelType
+  std::map<Models::ModelType, double> mu;  // ~= Probability of being in model ModelType
+
+  double alpha = 0.5;  // How much to weight old updates when getting average frequency
+  double relative_update_frequency = 0.5;
+  unsigned int updates_until_valid_track = 10;
+  double regularly_updated_frequncy_cutoff = 0.1;
```

Comment:
How should we do params outside the ROS nodes? Is there an easy way or do we need to come up with some custom solution?